### ISSUE TITLE
Clarify the behaviour of `roi` parameter for the `Refocus.autofocus` method

### ISSUE THREAD

**Issue by PinkShnack** [RI-imaging/nrefocus#17]
## (possible) Problem
I am a little confused by the `roi` parameter for the `Refocus.autofocus` method. The current docstring description is:

```markdown
roi: list or tuple or slice or ndarray
  Region of interest for which the metric will be minimized.
  This can be either a list [x1, y1, x2, y2], a tuple or
  list of slices or a numpy indexing array. If not given,
  the entire field will be used.
```

From `[x1, y1, x2, y2]` I would expect the x1, y1 to be the bottom left part of the image. However, in  `nrefocus.iface.base.Refocus.autofocus`, the rearranging of the list into slices does not match this assumption.

## Possible Solution
The roi parameter should be given in the numpy slicing order i.e. [top, bottom, left, right] which would be converted to the following slices: `[top:bottom, left:right]`.

## Related
Related to #12, which I agree should also be allowed if possible.

Related to PR #16

**Comment by paulmueller** [comment]
If I understand correctly, you would expect that the following would yield identical results?

- `roi = [x1, y1, x2, y2]`
- `roi = (slice(x1, x2), slice(y1, y2)`

But actually the results are what you would get with `roi = (slice(y1, y2), slice(x1, x2)`?

Maybe we should bury the concept of x and y and instead use the term `axis` in the docs (and in the code).

This is indeed confusing and the reason is probably (historically) related to how matplotlib plots data and how I defined ROIs in an earlier version of DryMass.

In any case, this is something that should be fixed (rather sooner than later) and unfortunately it will break things downstream in DryMass (and probably also qpimage).

**Comment by PinkShnack** [comment]
I would have expected `roi=[x1, y1, x2, y2]` to crop the 2d array like so: `crop = image[y2: y1, x1, y2]` because I assumed the origin was the bottom left of the image (y1 < y2).

I will make the changes in #16 if that is suitable?

**Comment by paulmueller** [comment]
This must be a typo, `image[y2: y1, x1: y2]` does not make sense at all!

**Comment by paulmueller** [comment]
But yes, please make the changes in #16 :+1: :pray: .

**Comment by PinkShnack** [comment]
I have the feeling the issue here is just my understanding of the `roi` param because it isn't described clearly enough. Then we can decide if we want to actually change what is expected from the user and if there is a bug in the code.

`[x1, y1, x2, y2]` for me means the coordinates on a standard coordinate grid:

![image](https://user-images.githubusercontent.com/43960046/196397645-5fc93d6f-07e4-4ec5-bad2-2f65ac1fa64c.png)


But we usually use numpy index notation in python where this grid (for the simple 2D situation):

![image](https://user-images.githubusercontent.com/43960046/196397726-dfe4400e-2379-46ee-8f5b-3ba4bbc920e9.png)


So we should tell the user to follow numpy indexing:

```python
[axis_0_start, axis_0_end, axis_1_start, axis_1_end]
# and within the function this will change to
[axis_0_start: axis_0_end, axis_1_start: axis_1_end]
```

Currently, the x axis is taken as the first slice, which is just incorrect as far as I know, and therefore a bug.

**Comment by paulmueller** [comment]
> > This must be a typo, `image[y2: y1, x1: y2]` does not make sense at all!
> 
> Indeed, I fixed it ([#17 (comment)](https://github.com/RI-imaging/nrefocus/issues/17#issuecomment-1278595516))

But you surely meant `crop = image[y1: y2, x1: x2]`, correct? Even more reason to move to "top, bottom, left, and right"...

**Comment by PinkShnack** [comment]
> > > This must be a typo, `image[y2: y1, x1: y2]` does not make sense at all!
> > 
> > 
> > Indeed, I fixed it ([#17 (comment)](https://github.com/RI-imaging/nrefocus/issues/17#issuecomment-1278595516))
> 
> But you surely meant `crop = image[y1: y2, x1: x2]`, correct? Even more reason to move to "top, bottom, left, and right"...

Nope, I meant it as `image[y2: y1, x1: y2]` because the question is **_where is y1_**. If we assume origin is bottom left then y1 is the smaller value, but that means the current slices go against numpy index notation.

**Comment by PinkShnack** [comment]
It could be that the solution here is only include the following code in the roi check:
```python
if all(isinstance(s, slice) for s in roi):
    # pass the slice directly
    pass
...  # the rest of the check
```

and to include in the docstring:
`Any list or tuples of slices will be passed directly as is in the expected numpy indexing order`

Because at the moment this is actually the behaviour! a list or tuple of slices just gets passed along.
With the above, the original behaviour is left alone, and the behaviour when using slices is clarified.

**Comment by PinkShnack** [comment]
I guess this can be closed due to #16 being merged?
